### PR TITLE
Have XFEM to automatically provide crack front points to CrackFrontDefinition 

### PR DESCRIPTION
**Pull request by jiangwen84**
closes ref #10265

1. add crack front points provider
2. xfem geometriccutter derives from it and provides crack front points to CrackFrontDefinition
3. modified the related DomainIntegralAction
4. added some tests.

### CODE REVIEW

**Comment by jiangwen84** [comment]
@bwspenc It is not quite completed yet. However, I would like you to take a look at the code design before I start the cleanup work.

File: modules/tensor_mechanics/src/userobjects/CrackFrontPointsProvider.C [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+/*             See LICENSE for full restrictions                */
+/****************************************************************/
+#include "CrackFrontPointsProvider.h"
+
+template <>
+InputParameters
+validParams<CrackFrontPointsProvider>()
+{
+  InputParameters params = validParams<GeneralUserObject>();
+  params.addClassDescription("Crack front points provider base class. Override the "
```
Comment (bwspenc) [inline modules/tensor_mechanics/src/userobjects/CrackFrontPointsProvider.C]: Maybe reword this description as:
`Base class for a class that can provide a set of points along a crack front. The virtual functions must be overridden by a derived class to provide this functionality.`

File: modules/tensor_mechanics/src/userobjects/CrackFrontDefinition.C [hunk 2]
```diff
@@ -81,6 +81,12 @@ addCrackFrontDefinitionParams(InputParameters & params)
                              q_function_type,
                              "The method used to define the integration domain. Options are: " +
                                  q_function_type.getRawNames());
+  params.addParam<UserObjectName>(
+      "crack_front_points_privider",
```
Comment (bwspenc) [inline modules/tensor_mechanics/src/userobjects/CrackFrontDefinition.C]: privider->provider

Comment (bwspenc) [reply modules/tensor_mechanics/src/userobjects/CrackFrontDefinition.C]: Even though the name of this base class is CrackFrontPointProvider, in practice, I think the only way we're going to use it is if it is an XFEM GeometricCutUserObject. I'm kind of torn about what the name of this parameter should be. Leave it as-is, but we may want to change that.

File: modules/tensor_mechanics/src/userobjects/CrackFrontDefinition.C [hunk 3]
```diff
@@ -81,6 +81,12 @@ addCrackFrontDefinitionParams(InputParameters & params)
...
+      "The UserObject provides the crack front points from XFEM GeometricCutObject");
+  params.addParam<unsigned int>(
+      "number_crack_front_points",
+      "The number of crack front points, only needed if crack_front_points_provider is used.");
```
Comment (bwspenc) [inline modules/tensor_mechanics/src/userobjects/CrackFrontDefinition.C]: It seems like this might be better in the input for the CrackFrontPointProvider. It would clean up the code on the fracture integral side a bit. I guess it would need to be optional there, and you'd have to generate an error if the points were requested and it wasn't supplied.

File: modules/xfem/src/userobjects/EllipseCutUserObject.C [hunk 4]
```diff
@@ -91,3 +91,18 @@ EllipseCutUserObject::isInsideCutPlane(Point p) const
...
+  Point v2 = _unit_vec2;
+  _crack_front_points.resize(_number_crack_front_points);
+  // parametric circle in 3D: center + r * cos(theta) * v1 + r * sin(theta) * v2
+  for (unsigned int i = 0; i < _number_crack_front_points; ++i)
+  {
+    Real theta = 2.0 * libMesh::pi / _number_crack_front_points * i;
+    _crack_front_points[i] =
+        _center + _long_axis * std::cos(theta) * v1 + _short_axis * std::sin(theta) * v2;
+  }
+}
```
Comment (bwspenc) [inline modules/xfem/src/userobjects/EllipseCutUserObject.C]: It would be nice to have these uniformly spaced by arc length rather than by the angle on the parametric circle before we merge this, although it looks like this is a little bit messy. Looks like you need to use an elliptic integral. 
http://mathworld.wolfram.com/Ellipse.html
https://math.stackexchange.com/questions/172766/calculating-equidistant-points-around-an-ellipse-arc
There are a number of approximations of elliptic integrals out there that aren't too bad to implement.

To avoid holding this commit up, maybe we should just skip the ellipse for now while we sort that out.

File: modules/tensor_mechanics/include/userobjects/CrackFrontPointsProvider.h [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+  virtual void initialSetup() override{};
+  virtual void initialize() override{};
+  virtual void execute() override{};
+  virtual void finalize() override{};
+
+  /**
+   *
+   */
+  virtual void calculateCrackFrontPoints(){};
```
Comment (bwspenc) [inline modules/tensor_mechanics/include/userobjects/CrackFrontPointsProvider.h]: The implementation in the base class should throw a MooseError saying that you're trying to use this function for an object that doesn't override that method.

File: modules/tensor_mechanics/include/userobjects/CrackFrontPointsProvider.h [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+
+  /**
+   *
+   */
+  virtual const std::vector<Point> & getCrackFrontPoints() const { return _crack_front_points; };
+
+  /**
+   *
+   */
+  unsigned int num_crack_front_points() const { return _number_crack_front_points; };
```
Comment (bwspenc) [inline modules/tensor_mechanics/include/userobjects/CrackFrontPointsProvider.h]: Follow the camelcase naming convention (numCrackFrontPoints)

File: modules/tensor_mechanics/include/userobjects/CrackFrontPointsProvider.h [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+
+  /**
+   *
+   */
+  virtual void set_num_crack_front_points(unsigned int number_crack_front_points)
```
Comment (bwspenc) [inline modules/tensor_mechanics/include/userobjects/CrackFrontPointsProvider.h]: follow naming convention here too

File: modules/tensor_mechanics/include/userobjects/CrackFrontPointsProvider.h [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+
+  /**
+   *
+   */
+  virtual void calculateCrackFrontPoints(){};
+
+  /**
+   *
+   */
+  virtual const std::vector<Point> & getCrackFrontPoints() const { return _crack_front_points; };
```
Comment (bwspenc) [inline modules/tensor_mechanics/include/userobjects/CrackFrontPointsProvider.h]: I don't really see the need to have both getCrackFrontPoints and calculateCrackFrontPoints. I'd keep getCrackFrontPoints and get rid of the other one.

**Review by bwspenc — CHANGES_REQUESTED** [review]
Just a few minor issues. The basic architecture looks good to me.

**Comment by jiangwen84** [comment]
@bwspenc I addressed all the comments. Thanks.

**Comment by jiangwen84** [comment]
@bwspenc any more comments? Thanks.

File: modules/tensor_mechanics/include/userobjects/CrackFrontPointsProvider.h [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+  virtual void execute() override{};
+  virtual void finalize() override{};
+
+  /** get a set of points along a crack front from a XFEM GeometricCutUserObject
+   * @return A vector which contains all crack front points
+   */
+  virtual const std::vector<Point>
+  getCrackFrontPoints(unsigned int /*num_crack_front_points*/) const
+  {
+    mooseError("You're trying to use getCrackFrontPoints() for an object that doesn't "
```
Comment (permcody) [inline modules/tensor_mechanics/include/userobjects/CrackFrontPointsProvider.h]: If you are going to `mooseError` here, did you consider an abstract method instead? That changes a runtime error into a compile time error unless you have objects out there that you need to support. I'd recommend the change unless you have a specific reason not to.

File: modules/tensor_mechanics/include/userobjects/CrackFrontPointsProvider.h [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+ */
+class CrackFrontPointsProvider : public GeneralUserObject
+{
+public:
+  CrackFrontPointsProvider(const InputParameters & parameters);
+
+  virtual void initialSetup() override{};
+  virtual void initialize() override{};
+  virtual void execute() override{};
+  virtual void finalize() override{};
```
Comment (permcody) [inline modules/tensor_mechanics/include/userobjects/CrackFrontPointsProvider.h]: You are aren't using any part of the GeneralUserObject interface. Should this just be a new base class instead?

File: modules/xfem/src/userobjects/CircleCutUserObject.C [hunk 11]
```diff
@@ -67,3 +67,22 @@ CircleCutUserObject::isInsideCutPlane(Point p) const
...
 }
+
+const std::vector<Point>
+CircleCutUserObject::getCrackFrontPoints(unsigned int number_crack_front_points) const
+{
+  std::vector<Point> crack_front_points(number_crack_front_points);
+  Point v1 = _vertices[0] - _center;
+  Xfem::normalizePoint(v1);
+  Point v2 = _normal.cross(v1);
+  Xfem::normalizePoint(v2);
```
Comment (permcody) [inline modules/xfem/src/userobjects/CircleCutUserObject.C]: v2.norm() - This already exists as a member.

File: modules/tensor_mechanics/src/actions/DomainIntegralAction.C [hunk 12]
```diff
@@ -84,6 +84,7 @@ validParams<DomainIntegralAction>()
 DomainIntegralAction::DomainIntegralAction(const InputParameters & params)
   : Action(params),
     _boundary_names(getParam<std::vector<BoundaryName>>("boundary")),
+    _use_crack_front_points_provider(false),
```
Comment (permcody) [inline modules/tensor_mechanics/src/actions/DomainIntegralAction.C]: You added a new unsigned int but didn't initialize it in all branches in the constructor. That may not bite you today, but a little refactor and you could be accessing uninitialized values.

**Review by permcody — COMMENTED** [review]
I'm not completely up on what you are trying to do, or how you are working it in while supporting existing methods, but I'm not sure I like where you are going here. Your new object, doesn't use _any_ of the UserObject interface. It's not the first object to do this but it begs the question, are UserObjects the right way to implement this? Why not just create a new base class?

If you stick with this design, I have a few other code comments to help your write more defensive code.

File: modules/tensor_mechanics/include/userobjects/CrackFrontPointsProvider.h (hunk 10 above)
Comment (jiangwen84) [reply modules/tensor_mechanics/include/userobjects/CrackFrontPointsProvider.h]: CrackFrontDefinition is in TensorMechanics and we want it to use with XFEMCutterUO which exists in XFEM. Now XFEM module depends on TensorMechanics module and we think to have a base class in TensorMechanics and let XFEMCutterUO to derive from it would resolve a dependence issue. This new base class should be UO such that we can use getUserObject interface. Do you have better suggestion?

File: modules/tensor_mechanics/src/actions/DomainIntegralAction.C (hunk 12 above)
Comment (jiangwen84) [reply modules/tensor_mechanics/src/actions/DomainIntegralAction.C]: are you talking about _use_crack_front_points_provider? I think I have initialized it in the Constructor.

File: modules/xfem/src/userobjects/CircleCutUserObject.C (hunk 11 above)
Comment (jiangwen84) [reply modules/xfem/src/userobjects/CircleCutUserObject.C]: I will use it. Thanks.

File: modules/tensor_mechanics/include/userobjects/CrackFrontPointsProvider.h (hunk 10 above)
Comment (permcody) [reply modules/tensor_mechanics/include/userobjects/CrackFrontPointsProvider.h]: The first part of this comment doesn't make sense to me. Are you trying to avoid depending on tensor mechanics? If you want to make that dependency permanent we can in the modules.mk file, then there's no problem with that.

I don't see any reason to use the getUserObject interface. You are just pulling this object into other objects but there's not execution dependencies. You're just using this object. From that standpoint you could just create a new base class and include the header any place it needs to be used and pull it in. I just don't see what the user object or the interface buys you since you are bypassing 100% of the interface anyway.

File: modules/tensor_mechanics/src/actions/DomainIntegralAction.C (hunk 12 above)
Comment (permcody) [reply modules/tensor_mechanics/src/actions/DomainIntegralAction.C]: No sorry, I'm talking about the "num crack front points" variable. This is just where that initialization should go. Alternatively, you can initialize it to zero (or some other sensible value) in all branches in your constructor body.

File: modules/tensor_mechanics/include/userobjects/CrackFrontPointsProvider.h (hunk 10 above)
Comment (jiangwen84) [reply modules/tensor_mechanics/include/userobjects/CrackFrontPointsProvider.h]: We want XFEM depend on TM. But the question is that CrackFrontDefinition implemented in TM need to use the XFEMCutterUO class which is in XFEM. This seems to have TM depend on XFEM. We do not have this dependence issue and I am not sure what is the best to do it.

File: modules/tensor_mechanics/include/userobjects/CrackFrontPointsProvider.h (hunk 9 above)
Comment (jiangwen84) [reply modules/tensor_mechanics/include/userobjects/CrackFrontPointsProvider.h]: Yes, I can make it pure virtual function. Thanks.

File: modules/tensor_mechanics/include/userobjects/CrackFrontPointsProvider.h (hunk 10 above)
Comment (permcody) [reply modules/tensor_mechanics/include/userobjects/CrackFrontPointsProvider.h]: Oh I see, Hmm... technically that would work on the compilation side of things. I'm not sure I really want it logically though. It would mean that people get XFEM if they use TM too.

Comment (bwspenc) [reply modules/tensor_mechanics/include/userobjects/CrackFrontPointsProvider.h]: We designed it this way specifically because we did not want to force people using TM to have to pull in XFEM.

Comment (jiangwen84) [reply modules/tensor_mechanics/include/userobjects/CrackFrontPointsProvider.h]: At beginning, all the XFEMCutterObjects were not UO and we ended up with creating them and providing their parameters in XFEMAction, which is a bit cumbersome. And then we decide to have XFEMCutterObject be generalUO and we could provide parameters easier and add/remove them in the input file. The XFEMCutterObject does not really override those basic virtual functions. The new CrackFrontPointsProvider UO here just provides a base class to let XFEMCutterObject derive from. And I could also use getUserObject in the CrackFrontDefinition class to get it.

Comment (bwspenc) [reply modules/tensor_mechanics/include/userobjects/CrackFrontPointsProvider.h]: I talked this over with @permcody.  The main reason for using UserObjects without using any of that class's interface is that it is convenient to create and parse the parameters for these objects. Until we have something that offers us that for a more stripped down object, let's keep using UserObjects for this purpose.

**Comment by jiangwen84** [comment]
I made several minor changes according to Cody's suggestions. It is ready for another review.

File: modules/tensor_mechanics/src/actions/DomainIntegralAction.C [hunk 13]
```diff
@@ -774,6 +787,8 @@ DomainIntegralAction::calcNumCrackFrontPoints()
   }
   else if (_crack_front_points.size() != 0)
     num_points = _crack_front_points.size();
+  else if (_use_crack_front_points_provider)
+    num_points = getParam<unsigned int>("number_crack_front_points");
```
Comment (bwspenc) [inline modules/tensor_mechanics/src/actions/DomainIntegralAction.C]: Can't you just use `_number_crack_front_points`?

File: modules/tensor_mechanics/src/actions/DomainIntegralAction.C [hunk 14]
```diff
@@ -136,6 +138,12 @@ DomainIntegralAction::DomainIntegralAction(const InputParameters & params)
   {
     _crack_front_points = getParam<std::vector<Point>>("crack_front_points");
   }
+  if (isParamValid("crack_front_points_provider") && isParamValid("number_crack_front_points"))
+  {
+    _use_crack_front_points_provider = true;
+    _crack_front_points_provider = getParam<UserObjectName>("crack_front_points_provider");
+    _number_crack_front_points = getParam<unsigned int>("number_crack_front_points");
+  }
```
Comment (bwspenc) [inline modules/tensor_mechanics/src/actions/DomainIntegralAction.C]: I think you need an error check for the case where 'crack_front_points_provider' is supplied, but 'number_crack_front_points' isn't.

File: modules/tensor_mechanics/include/userobjects/CrackFrontDefinition.h [hunk 15]
```diff
@@ -162,6 +163,8 @@ class CrackFrontDefinition : public GeneralUserObject, public BoundaryRestrictab
   std::vector<bool> _is_point_on_intersecting_boundary;
   std::vector<Real> _j_integral_radius_inner;
   std::vector<Real> _j_integral_radius_outer;
+  const CrackFrontPointsProvider * _crack_front_points_provider;
+  unsigned int _number_crack_front_points;
```
Comment (bwspenc) [inline modules/tensor_mechanics/include/userobjects/CrackFrontDefinition.h]: The name `_number_crack_front_points` could be confusing because it is specifically tied to the use of `crack_front_points_provider`.  Maybe something like `_num_points_from_provider` would be more clear.

File: modules/tensor_mechanics/src/userobjects/CrackFrontDefinition.C [hunk 16]
```diff
@@ -81,6 +81,12 @@ addCrackFrontDefinitionParams(InputParameters & params)
                              q_function_type,
                              "The method used to define the integration domain. Options are: " +
                                  q_function_type.getRawNames());
+  params.addParam<UserObjectName>(
+      "crack_front_points_provider",
+      "The UserObject provides the crack front points from XFEM GeometricCutObject");
+  params.addParam<unsigned int>(
+      "number_crack_front_points",
```
Comment (bwspenc) [inline modules/tensor_mechanics/src/userobjects/CrackFrontDefinition.C]: Same deal here -- maybe this should be changed to something like 'number_points_from_provider' to make it clear that this applies specifically to the case where you're using a provider.

File: modules/tensor_mechanics/src/actions/DomainIntegralAction.C (hunk 14 above)
Comment (bwspenc) [reply modules/tensor_mechanics/src/actions/DomainIntegralAction.C]: Actually, maybe that's covered by your error check in CrackFrontDefinition.

File: modules/tensor_mechanics/src/userobjects/CrackFrontDefinition.C [hunk 17]
```diff
@@ -117,7 +137,8 @@ CrackFrontDefinition::CrackFrontDefinition(const InputParameters & parameters)
                  "set by user!");
   }
   else
-    mooseError("In CrackFrontDefinition, must define either 'boundary' or 'crack_front_points'");
+    mooseError("In CrackFrontDefinition, must define as one of 'boundary', 'crack_front_points' "
```
Comment (bwspenc) [inline modules/tensor_mechanics/src/userobjects/CrackFrontDefinition.C]: Remove 'as'

File: modules/tensor_mechanics/src/userobjects/CrackFrontDefinition.C [hunk 18]
```diff
@@ -98,17 +104,31 @@ CrackFrontDefinition::CrackFrontDefinition(const InputParameters & parameters)
...
     _t_stress(getParam<bool>("t_stress")),
     _q_function_rings(getParam<bool>("q_function_rings")),
-    _q_function_type(getParam<MooseEnum>("q_function_type"))
+    _q_function_type(getParam<MooseEnum>("q_function_type")),
+    _crack_front_points_provider(nullptr)
 {
   if (isParamValid("crack_front_points"))
   {
+    if (isParamValid("crack_front_points_provider"))
+      mooseError("As crack_front_points have been provided, the crack_front_points_provider will "
```
Comment (bwspenc) [inline modules/tensor_mechanics/src/userobjects/CrackFrontDefinition.C]: I guess we need to add a check for the case where 'boundary' is specified too.

File: modules/tensor_mechanics/src/userobjects/CrackFrontDefinition.C [hunk 19]
```diff
@@ -98,17 +104,31 @@ CrackFrontDefinition::CrackFrontDefinition(const InputParameters & parameters)
...
+                 "not be used and needs to be removed.");
     _crack_front_points = getParam<std::vector<Point>>("crack_front_points");
     _geom_definition_method = CRACK_FRONT_POINTS;
     if (_t_stress)
       mooseError("t_stress not yet supported with crack_front_points");
     if (_q_function_rings)
       mooseError("q_function_rings not supported with crack_front_points");
   }
+  else if (isParamValid("crack_front_points_provider"))
+  {
```
Comment (bwspenc) [inline modules/tensor_mechanics/src/userobjects/CrackFrontDefinition.C]: Check to make sure 'boundary' is not specified here.

**Review by bwspenc — CHANGES_REQUESTED** [review]
close...

**Comment by jiangwen84** [comment]
@bwspenc  All the comments are addressed. Thanks.

File: modules/tensor_mechanics/include/actions/DomainIntegralAction.h [hunk 20]
```diff
@@ -49,6 +49,9 @@ class DomainIntegralAction : public Action
   std::set<INTEGRAL> _integrals;
   const std::vector<BoundaryName> & _boundary_names;
   std::vector<Point> _crack_front_points;
+  UserObjectName _crack_front_points_provider;
+  unsigned int _number_crack_front_points;
```
Comment (bwspenc) [inline modules/tensor_mechanics/include/actions/DomainIntegralAction.h]: Just one tiny thing: it would better if this were named `_num_points_from_provider` as it is in CrackFrontDefinition. You could actually just eliminate this as a member variable in this class, and grab it from the parameters when you need it.

**Review by bwspenc — APPROVED** [review]
I don't think my one small comment is worth holding up this commit for.  I'm going ahead and merging this now, but if you could address that when you're in this code again, that would be great.

File: modules/tensor_mechanics/include/actions/DomainIntegralAction.h (hunk 20 above)
Comment (jiangwen84) [reply modules/tensor_mechanics/include/actions/DomainIntegralAction.h]: Yes, I will make the change when I fix the closed_loop issue. Thanks.
